### PR TITLE
Add ability for plugins to handle file change notifications

### PR DESCRIPTION
**Pull request by pepeiborra**
Rebased #1585 and extracted ghcide notification handlers

### CODE REVIEW

File: hls-plugin-api/src/Ide/Types.hs [hunk 1]
```diff
@@ -180,18 +181,74 @@ instance PluginMethod TextDocumentRangeFormatting where
...
+--   = PluginNotificationHandler (PluginNotificationMethodHandler a m)`
+
+{-
+From Zubin
+alanz_: I would say `newtype PluginNotificationHandler a (m :: Method FromClient Notification) = PluginNotificationHandler (PluginNotificationMethodHandler a m)`
+16:28 and `newtype PluginNotificationHandlers a = PluginNotificationHandlers (DMap SMethod (PluginNotificationHandler a))
+
+-}
+
+newtype PluginHandlers a             = PluginHandlers             (DMap IdeMethod       (PluginHandler a))
```
Comment (alanz) [inline hls-plugin-api/src/Ide/Types.hs:249]: This section from here down is a repeat of what is in  ghcide/src/Development/IDE/Plugin/HLS.hs

**Review by alanz — COMMENTED** [review]
There seems to be some duplication between the HLS and ghcide side.

Otherwise seems to be what I had before.

File: hls-plugin-api/src/Ide/Types.hs [hunk 2]
```diff
@@ -180,18 +181,74 @@ instance PluginMethod TextDocumentRangeFormatting where
...
+instance GCompare IdeNotification where
+  gcompare (IdeNotification a) (IdeNotification b) = gcompare a b
+
 -- | Combine handlers for the
 newtype PluginHandler a (m :: Method FromClient Request)
   = PluginHandler (PluginId -> a -> MessageParams m -> LspM Config (NonEmpty (Either ResponseError (ResponseResult m))))
 
-newtype PluginHandlers a = PluginHandlers (DMap IdeMethod (PluginHandler a))
+newtype PluginNotificationHandler a (m :: Method FromClient Notification)
+  = PluginNotificationHandler (PluginId -> a -> MessageParams m -> LspM Config (NonEmpty ()))
```
Comment (wz1000) [inline hls-plugin-api/src/Ide/Types.hs:238]:
```suggestion
  = PluginNotificationHandler (PluginId -> a -> MessageParams m -> LspM Config ())
```

File: hls-plugin-api/src/Ide/Types.hs (hunk 1 above)
Comment (pepeiborra) [reply hls-plugin-api/src/Ide/Types.hs:249]: Which one is redundant?

Comment (pepeiborra) [reply hls-plugin-api/src/Ide/Types.hs:249]: Just to clarify, I haven't added anything. If there is any duplication, it was already present in your branch.

Comment (alanz) [reply hls-plugin-api/src/Ide/Types.hs:249]: fair enough :)

**Comment by isovector** [comment]
Is this capable of watching arbitrary files on the filesystem? I wanted such a thing in #524, in order to watch `$XDG_CONFIG/wingman`.

**Comment by pepeiborra** [comment]
> Is this capable of watching arbitrary directories on the filesystem? I wanted such a thing in #524, in order to watch `$XDG_CONFIG/wingman`. 

No, only workspace folders. Outside the workspace, you would need to implement your own watching

**Comment by jneira** [comment]
> 
> 
> > Is this capable of watching arbitrary directories on the filesystem? I wanted such a thing in #524, in order to watch `$XDG_CONFIG/wingman`.
> 
> No, only workspace folders. Outside the workspace, you would need to implement your own watching

I would need it for hlint config files for #594, maybe it worths to implement it in some common place (hls-plugin-api?)

**Comment by pepeiborra** [comment]
> > > Is this capable of watching arbitrary directories on the filesystem? I wanted such a thing in #524, in order to watch `$XDG_CONFIG/wingman`.
> > 
> > 
> > No, only workspace folders. Outside the workspace, you would need to implement your own watching
> 
> I would need it for hlint config files for #594, maybe it worths to implement it in some common place (hls-plugin-api?)

Rather than turning HLS into an operating system, I would push back on these feature requests and ask people to reload

**Review by alanz — APPROVED** [review]

File: hls-plugin-api/src/Ide/Types.hs [hunk 3]
```diff
@@ -191,19 +192,64 @@ instance PluginMethod TextDocumentRangeFormatting where
   pluginEnabled _ pid conf = (PluginId $ formattingProvider conf) == pid
   combineResponses _ _ _ _ (x :| _) = x
 
+-- ---------------------------------------------------------------------
+
+-- | Notifications that can be handled by plugins.
+-- 'ExtraParams' captures any extra data the IDE passes to the handlers for this method
+class HasTracing (MessageParams m) => PluginNotification m where
```
Comment (wz1000) [inline hls-plugin-api/src/Ide/Types.hs:199]: I don't think we really need this class. Isn't `HasTracing (MessageParams m)` sufficient?

Comment (wz1000) [reply hls-plugin-api/src/Ide/Types.hs:199]: It doesn't really make sense to disable notifications for a particular plugin via the config I think. Its more of an internal implementation detail that shouldn't be visible to the user.

We could disable notifications from a plugin if the plugin itself is disabled.

Comment (pepeiborra) [reply hls-plugin-api/src/Ide/Types.hs:199]: Isn't this class also used to restrict what `Method`s are allowed to be used in `IdeNotification`?

Comment (wz1000) [reply hls-plugin-api/src/Ide/Types.hs:199]: Yes, but is there any reason to have such a restriction?

Comment (pepeiborra) [reply hls-plugin-api/src/Ide/Types.hs:199]: I'll remove it

**Review by wz1000 — APPROVED** [review]

**Pull request merged**